### PR TITLE
Make conditions work properly for every case

### PR DESCRIPTION
**Pull request by uniumuniu**
- Refactor Conditions
- Add unit tests for Conditions
- Add dependencies for semver comparison and murmurhash

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -57,12 +59,14 @@ dependencies {
 
     // This dependency is used internally, and not exposed to consumers on their own compile classpath.
     implementation("com.google.guava:guava:31.1-jre")
+    implementation("net.swiftzer.semver:semver:1.3.0")
+    implementation("com.goncalossilva:murmurhash:0.4.0")
```
Comment (fahad19) [inline build.gradle.kts:63]: 💖

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt [hunk 2]
```diff
@@ -1,84 +1,134 @@
...
+import com.featurevisor.types.Operator.NOT_CONTAINS
+import com.featurevisor.types.Operator.NOT_EQUALS
+import com.featurevisor.types.Operator.NOT_IN_ARRAY
+import com.featurevisor.types.Operator.SEMVER_EQUALS
+import com.featurevisor.types.Operator.SEMVER_GREATER_THAN
+import com.featurevisor.types.Operator.SEMVER_GREATER_THAN_OR_EQUAL
+import com.featurevisor.types.Operator.SEMVER_LESS_THAN
+import com.featurevisor.types.Operator.SEMVER_LESS_THAN_OR_EQUAL
+import com.featurevisor.types.Operator.SEMVER_NOT_EQUALS
+import com.featurevisor.types.Operator.STARTS_WITH
```
Comment (fahad19) [inline src/main/kotlin/com/featurevisor/sdk/Conditions.kt:30]: that's a loooot of imports :-o

is it more idiomatic in Kotlin side of the world to have separate files per type?

**Review by fahad19 — APPROVED** [review]
🚀

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt (hunk 2 above)
Comment (uniumuniu) [reply src/main/kotlin/com/featurevisor/sdk/Conditions.kt:30]: The short answer is yes, but it really depends. But let's find some middle ground for this project.

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt [hunk 3]
```diff
@@ -1,84 +1,134 @@
...
-                    Operator.equals -> return contextValue.value == conditionValue.value
-                    Operator.notEquals -> return contextValue.value != conditionValue.value
-                    Operator.contains -> return contextValue.value.contains(conditionValue.value)
-                    Operator.notContains ->
-                        return !contextValue.value.contains(conditionValue.value)
+                    EQUALS -> attributeValue.value == conditionValue.value
+                    NOT_EQUALS -> attributeValue.value != conditionValue.value
+                    CONTAINS -> attributeValue.value.contains(conditionValue.value)
+                    NOT_CONTAINS ->
+                        attributeValue.value.contains(conditionValue.value).not()
```
Comment (mazabin) [inline src/main/kotlin/com/featurevisor/sdk/Conditions.kt:45]: Separate line on purpose? Looks like it should fit in one.

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt [hunk 4]
```diff
@@ -1,84 +1,134 @@
...
 
-                    Operator.startsWith ->
-                        return contextValue.value.startsWith(conditionValue.value)
+                    STARTS_WITH ->
+                        attributeValue.value.startsWith(conditionValue.value)
```
Comment (mazabin) [inline src/main/kotlin/com/featurevisor/sdk/Conditions.kt:48]: As above.

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt (hunk 3 above)
Comment (uniumuniu) [reply src/main/kotlin/com/featurevisor/sdk/Conditions.kt]: Fixed formatting

File: src/main/kotlin/com/featurevisor/sdk/Conditions.kt [hunk 5]
```diff
@@ -1,84 +1,134 @@
...
+    }
 
-        // boolean
-        if (contextValue is AttributeValue.BooleanValue && value is ConditionValue.BooleanValue) {
-            // boolean / boolean
-            when (operator) {
-                Operator.equals -> return contextValue.value == value.value
-                Operator.notEquals -> return contextValue.value != value.value
-                else -> return false
+    fun allConditionsAreMatched(condition: Condition, context: Context): Boolean {
```
Comment (mazabin) [inline src/main/kotlin/com/featurevisor/sdk/Conditions.kt:110]: Could use expression body - I don't remember what's in codestyle here, but next method `compareVersions` uses that, still with return type added.

File: src/test/kotlin/com/featurevisor/sdk/ConditionsTest.kt [hunk 6]
```diff
@@ -1,112 +1,513 @@
...
-    fun testEqualsOperatorForStrings() {
+    fun `EQUALS operator works for strings`() {
         val condition =
-            PlainCondition(
-                "browser_type",
-                Operator.equals,
-                ConditionValue.StringValue("chrome")
+            Condition.Plain(
+                attributeKey = "browser_type",
+                operator = Operator.EQUALS,
```
Comment (mazabin) [inline src/test/kotlin/com/featurevisor/sdk/ConditionsTest.kt:17]: Static import on the Operator?

**Review by mazabin — APPROVED** [review]

**Pull request merged**